src/MongoDB.Driver.Core/Core/WireProtocol/Messages/CompressedMessage.cs

```diff
@@ -0,0 +1,75 @@
+﻿/* Copyright 2019-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using MongoDB.Bson.IO;
+using MongoDB.Driver.Core.Compression;
+using MongoDB.Driver.Core.Misc;
+using MongoDB.Driver.Core.WireProtocol.Messages.Encoders;
+
+namespace MongoDB.Driver.Core.WireProtocol.Messages
+{
+    /// <summary>
+    /// Represents a compressed message.
+    /// </summary>
+    public class CompressedMessage : MongoDBMessage
+    {
+        private readonly BsonStream _originalMessageBytes;
+        private readonly MongoDBMessage _originalMessage;
+        private readonly CompressorType _compressorType;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CompressedMessage"/> class.
+        /// </summary>
+        /// <param name="originalMessage">The original message.</param>
+        /// <param name="originalBytes">The original bytes.</param>
+        /// <param name="compressorType">The compressor type.</param>
+        public CompressedMessage(
+            MongoDBMessage originalMessage,
+            BsonStream originalBytes,
```

Comment:
We can replace `BsonStream` here on something like `ICompressedMessageContent` which is a new possible interface. And this interface will have different implementations for `Binary` and `Json` encoders. In the current implementation, this value is null in `JsonEncoder`

Reply:
Let's not be complicated.

JsonEncoders would not actually do any compression. They are only for debugging and logging, and their purpose is to provide a string representation of the message.